R/session_run_hooks_custom.R

```diff
@@ -1,134 +1,226 @@
+      losses <- values$results[["losses"]]
+      loss <- losses[[length(losses)]]
+      tokens <- list(loss = format(round(loss, 2), nsmall = 2), step = .n + 1)
+      .bar$tick(len = n, tokens = tokens)
     }
     
-  }
-
-  write_run_properties <- function(props) {
-    properties <- list()
-    properties$steps <- steps
-    properties$model <- props$model
-    tfruns::write_run_metadata("properties", properties)
+    session_run_hook(
+      
+      before_run = function(context) {
+        session_run_args(
+          losses = run_context_losses(context)
+        )
+      },
+      
+      after_run = function(context, values) {
+        
+        # update progress bar
+        update_progress(values)
+        
+        # save and update state
+        .values <<- values
+        .n <<- .n + 1
+      },
+      
+      end = function(session) {
+        
+        # if we ran as many steps as expected, bail
+        if (is.null(steps) || identical(.n, steps))
+          return()
+        
+        update_progress(.values, steps - .n)
+      }
+      
+    )
   }
   
-  session_run_hook(
-    before_run = function(context) write_run_properties(props),
-    after_run = function(context, values) on_metrics(FALSE),
-    end = function(session) on_metrics(TRUE)
-  )
+  list(hook_fn = hook_fn, type = "hook_progress_bar")
 }
 
-hook_progress_bar <- function(label, steps) {
-  
-  format <- if (is.null(steps))
-    paste("[:spin]", label, "-- loss: :loss, step: :step")
-  else
-    paste(label, ":current/:total [:bar] - ETA: :eta - loss: :loss")
-  
-  .values <- NULL
-  .n <- 0
-  .bar <- progress::progress_bar$new(
-    format = format,
-    total = steps %||% 1E6,
-    complete = "=",
-    incomplete = ".",
-    clear = FALSE,
-    width = min(getOption("width"), 80),
-    stream = stdout(),
-    show_after = 0
-  )
-  
-  update_progress <- function(values, n = 1) {
-    losses <- values$results[["losses"]]
-    loss <- losses[[length(losses)]]
-    tokens <- list(loss = format(round(loss, 2), nsmall = 2), step = .n + 1)
-    .bar$tick(len = n, tokens = tokens)
+
+# NOTE: we need to pad with an extra row of data to signal to
+# the viewer that there is more data incoming. by returning a
+# metrics dataframe with no padding, we signal to the viewer
+# that there is no more data incoming
+get_metrics_df <- function(mode_key, finalize = TRUE, steps = NULL) {
+  df <- as.data.frame(.globals$history[[mode_key]]$losses)
+  if (finalize)
+    return(df)
+  pad(df, steps %||% nrow(df) + 1)
+}
+
+#' Visualize the training or evaluation metrics
+#' 
+#' This function allows users to visualize the metrics produced in training or evaluation phases.
+#' 
+#' @param mode_key The mode when the metrics were collected that you want to visualize, e.g. "train"
+#' 
+#' @return The directory used to save the metrics metadata and generated html file.
+#' @export
+visualize_metrics <- function(mode_key = c("train", "eval")) {
```

Comment:
I think we should get rid of this function as it's only a temporary hack until we figure out why `hook_view_metrics doesn't work). We can just remove `hook_view_metrics()` entirely until we sort all of that out.

Reply:
Do you mean removing visualize_metrics() as well?

Reply:
Yes, I don't think we want to support that longer term so better we leave it out for now.

Reply:
Removed and added a function `get_saved_history` in https://github.com/rstudio/tfestimators/pull/111/commits/1be2d5593b1f3043e3b7b5e855dc1c5d46d73ad2.